src/DotNet.Testcontainers/Containers/ResourceReaper.cs

```diff
@@ -0,0 +1,288 @@
+    /// <returns>Task that completes when the <see cref="ResourceReaper" /> has been started.</returns>
+    public static async Task<ResourceReaper> GetAndStartNewAsync(Guid sessionId, CancellationToken ct = default)
+    {
+      var resourceReaper = new ResourceReaper(sessionId);
+
+      try
+      {
+        await resourceReaper.StartAsync(ct)
+          .ConfigureAwait(false);
+        return resourceReaper;
+      }
+      catch (Exception)
+      {
+        await resourceReaper.DisposeAsync()
+          .ConfigureAwait(false);
+        throw;
+      }
+    }
+
+    public async ValueTask DisposeAsync()
+    {
+      if (this.disposed)
+      {
+        return;
+      }
+
+      this.disposed = true;
+
+      this.cancellationTokenSource.Cancel();
+
+      if (this.connectionTask != null)
+      {
+        try
+        {
+          // Ensure that connectionTask stops eventually before ResourceReaper is considered as disposed
+          // If we remove this we need to consider that some tests don't fail if the connection Task is not stopped.
+          await this.connectionTask;
+        }
+        catch (Exception)
+        {
+          // ignore
+        }
+      }
+
+      this.cancellationTokenSource.Dispose();
+
+      if (this.resourceReaperContainer != null)
+      {
+        await this.resourceReaperContainer.DisposeAsync();
+      }
+    }
+
+    private async Task StartAsync(CancellationToken ct = default)
+    {
+      var ryukInitializedTaskSource = new TaskCompletionSource<bool>();
+
+      await this.resourceReaperContainer.StartAsync(ct)
+        .ConfigureAwait(false);
+
+      this.connectionTask = this.MaintainRyukConnection(ryukInitializedTaskSource);
+
+      await Task.WhenAny(
+        Task.Delay(Timeout.InfiniteTimeSpan, ct),
+        ryukInitializedTaskSource.Task
+      ).ConfigureAwait(false);
+
+      if (!ryukInitializedTaskSource.Task.IsCompleted)
+      {
+        throw new OperationCanceledException("Initializing ResourceReaper connection was cancelled!", ct);
+      }
+
+      if (ryukInitializedTaskSource.Task.IsFaulted)
+      {
+        throw ryukInitializedTaskSource.Task.Exception.InnerException;
+      }
+    }
+
+    private async Task MaintainRyukConnection(TaskCompletionSource<bool> ryukInitializedTaskSource)
+    {
+      using (var initTimeoutCancellationTokenSource = new CancellationTokenSource(InitTimeout))
+      {
+        using (var linkedCancellationTokenSource = CancellationTokenSource.CreateLinkedTokenSource(
+                 this.cancellationTokenSource.Token,
+                 initTimeoutCancellationTokenSource.Token))
+        {
+          var endpoint = new IPEndPoint(IPAddress.Loopback, this.resourceReaperContainer.GetMappedPublicPort(RyukPort));
+
+          while (!this.cancellationTokenSource.Token.IsCancellationRequested && (ryukInitializedTaskSource.Task.IsCompleted || !initTimeoutCancellationTokenSource.IsCancellationRequested))
```

Comment:
I think `!linkedCancellationTokenSource.IsCancellationRequested` is enough. In which case would it be not? I even think `if (!ryukInitializedTaskSource.Task.IsCompleted)` L:225 is unnecessary. In which case the while will fail and we set the task result?

Reply:
`!linkedCancellationTokenSource.IsCancellationRequested` would cancel the `MaintainRyukConnection` loop as soon as the `InitTimeout` has elapsed even if initialization was already successful.

`if (!ryukInitializedTaskSource.Task.IsCompleted)` is probably not necessary. It would only lead to registering the session id label again when the connection to ryuk was accidentally lost after initialization. Registering the label again is not necessary but also no problem.

Reply:
> `!linkedCancellationTokenSource.IsCancellationRequested` would cancel the `MaintainRyukConnection` loop as soon as the `InitTimeout` has elapsed even if initialization was already successful.

That's something I already was wondering about. In that case, we are breaking out of the second while loop, right? I checked a couple of times for an `OperationCanceledException`, but I didn't catch anything.

Doesn't the connection to Ryuk have to remain open? Do need to protect the second while loop from breaking?

Reply:
Yes the Ryuk connection should remain open, but after disconnect Ryuk waits 10 seconds for a reconnect. If that does not happen it triggers the clean up. I think it is only to be resilient against network issues.

Reply:
Well, why don't we remove the timeout and replace it with something like a circuit breaker? I think the timeout (the way we're using it right now) is a design flaw. It gets canceled even if the connection is established.

Reply:
I am not sure how this would look like. Would you try to implement it? As mentioned in another comment I have moved the timeout logic now into the `GetAndStartNewAsync` method.

Reply:
Should I try to add the circuit breaker on the weekend too? Or should we keep the timeout?

Reply:
I would be ok with the timeout. Maybe the circuit breaker could be introduced in a separate PR.